sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -0,0 +1,256 @@
+    }
+
+    fn stop(&self) -> Result<()> {
+        log::info!("Stopping OpenOCD...");
+
+        let Some(mut child) = self.openocd_server_process.take() else {
+            // OpenOCD wasn't started, do nothing
+            return Ok(());
+        };
+
+        // TODO: send "shutdown" command and wait before killing the process
+        Ok(child.kill()?)
+    }
+
+    /// Send a TCL command to OpenOCD and wait for its response.
+    fn send_tcl_cmd(&self, cmd: &str) -> Result<String> {
+        // Take the stream. The stream will not be replaced on communication
+        // errors, causing future commands to also fail.
+        let Some(mut stream) = self.openocd_socket_stream.take() else {
+            return Err(OpenOcdServerError::OpenOcdNotRunning.into());
+        };
+
+        // The protocol is to send the command followed by a `0x1a` byte,
+        // see https://openocd.org/doc/html/Tcl-Scripting-API.html#Tcl-RPC-server
+        stream
+            .write_all(cmd.as_bytes())
+            .context("failed to send a command to OpenOCD server")?;
+        stream
+            .write_all(&[0x1a])
+            .context("failed to send a command to OpenOCD server")?;
+
+        stream.flush().context("failed to flush stream")?;
+
+        let answer = stream
+            .borrow_mut()
+            .bytes()
+            .take_while(|x| match x {
+                Ok(x) => *x != 0x1a,
+                _ => false,
+            })
+            .collect::<Result<_, _>>()
+            .context("failed to read the result of a command from OpenOCD server")?;
+
+        self.openocd_socket_stream.replace(Some(stream));
+        String::from_utf8(answer).context("failed to parse OpenOCD response as UTF-8")
+    }
+}
+
+impl Drop for OpenOcdServer {
+    fn drop(&mut self) {
+        self.stop().ok();
+    }
+}
+
+impl Jtag for OpenOcdServer {
+    fn connect(&self, tap: JtagTap) -> Result<()> {
+        self.start(tap)
+    }
+
+    fn disconnect(&self) -> Result<()> {
+        self.stop()
+    }
+
+    fn read_lc_ctrl_reg(&self, reg: &LcCtrlReg) -> Result<u32> {
+        let reg_offset = reg.word_offset();
+        let cmd = format!("riscv dmi_read 0x{reg_offset:x}");
+        let response = self.send_tcl_cmd(cmd.as_str())?;
+
+        let value = parse_hex_word(response.trim()).context(format!(
+            "expected response to be hexadecimal word, got '{response}'"
+        ))?;
+
+        Ok(value)
+    }
+
+    fn write_lc_ctrl_reg(&self, reg: &LcCtrlReg, value: u32) -> Result<()> {
+        let reg_offset = reg.word_offset();
+        let cmd = format!("riscv dmi_write 0x{reg_offset:x} 0x{value:x}");
+        let response = self.send_tcl_cmd(cmd.as_str())?;
+
+        if !response.is_empty() {
+            bail!("unexpected response: '{response}'");
+        }
+
+        Ok(())
+    }
+
+    fn read_memory(&self, addr: u32, buf: &mut [u8]) -> Result<u32> {
+        let cmd = format!("read_memory 0x{addr:x} 8 {count}", count = buf.len());
```

Comment:
not sure this (and write function and halt/resumt) will work with LC_TAP selected. needs to be investigated. But lets do it in a followup

Reply:
ah you mentioned this in your comment above.